Remplace UserToken et Password par une gestion de Permission

thom4parisot opened the pull request:
- [x] supprime le code relatif à `UserToken`
- [x] supprime le code relatif à `Password`
- [x] ajuste le contenu du token JWT
- [x] les permissions sont résolues au niveau de l'API (au lieu du token)
- [x] ajout/suppression d'un user dans la route `/credentials` (permissions read/write)
- [x] ajout/suppression d'un user au niveau d'un article (permissions read/write)
- [x] migre les données
	- [x] User.authType
	- [x] Password.password ➡ User.password
	- [x] Password.users ➡ User.permissions (`UserPermission`)
	- [x] Article.owners ➡ Article.owner + Article.contributors (`ArticleContributor`)
---
- [x] le login par mot de passe fonctionne
- [x] la création d'un compte fonctionne
- [x] la modification d'un mot de passe fonctionne
- [x] les articles de l'utilisateur sont affichés
- [x] les articles partagés apparaissent dans `/articles`
- [x] je peux écrire dans un de mes articles
- [x] je peux écrire dans un article partagé
- [x] je peux prévisualiser un article
- [x] je peux prévisualiser un article qui m'a été partagé
- [x] je peux prévisualiser un article lisibile via un compte partagé
- [x] je ne peux pas partager un article qui m'a été partagé
- [x] je peux retirer une personne d'un partage de compte
- [x] je peux retirer une personne d'un partage d'articles

---

## Bugs repérés

- [x] sur la branche `master` je peux me logguer avec un user, tandis que sur cette branche la connexion m'est refusée
- [x] quand je recharge la page, je perds ma connexion

---

Pour lancer les migrations de base de données : 

```sh
# User/Passwords -> User/Permissions
docker cp graphql/data/20212114-passwords.js mongodb-stylo:/migration.js
docker-compose exec mongodb-stylo mongo stylo-prod /migration.js

# Article Owners -> Owner/Contributors
docker cp graphql/data/20220112-article-contributors.js mongodb-stylo:/migration.js
docker-compose exec mongodb-stylo mongo stylo-prod /migration.js
```

fixes #464
fixes #159
fixes #504 
refs #289

Commit:
Remove the Tokens concept

It was actually unusued in the codebase

Commit:
Remove Password concept

Commit:
Rename GraphQL Schema

Commit:
Use `gql` template string to build schema

Commit:
Local login works

Commit:
Password migration

thom4parisot commented:
Pour lancer la migration : 

```sh
docker-compose cp graphql/data/20212114-passwords.js mongodb-stylo:migration.js
docker-compose exec mongodb-stylo mongo migration.js
```

ggrossetie commented on graphql/schema.js:
```diff
@@ -0,0 +1,183 @@
+const gql = require('graphql-tag')
+const { buildASTSchema } = require('graphql')
+
+module.exports = buildASTSchema(gql`
```
Je ne vois pas de différence entre les deux approches... Est ce que je rate quelque chose ?

Ou du moins la définition semble être la même mais peut être que le fait d'utiliser `buildASTSchema` (plutôt que `builldSchema`) permet d'avoir accès à d'autres fonctionnalités ?

thom4parisot replied:
Oui, la coloration syntaxique dans ton IDE :-)

## Avec `gql`

<img width="1512" alt="Screenshot 2021-12-14 at 10 23 21" src="https://user-images.githubusercontent.com/138627/145970338-c456ca31-18a9-4486-8bc8-5df6da9d1359.png">

## Sans `gql`

<img width="1512" alt="image" src="https://user-images.githubusercontent.com/138627/145970424-1f2541da-8af2-4034-92e6-b94805530f72.png">

ggrossetie replied:
🤩

ggrossetie commented:
> Password.users :arrow_right: User.permissions (UserPermission)
> Article.users :arrow_right: Article.permissions (ArticlePermission)

Initialement j'avais pensé mettre toutes les permissions sur l'utilisateur mais peut être que c'est plus compliqué car on ne peut pas faire des jointures.
Par contre si on fait ça, la notion de scope perd de son intérêt non ?

Mon idée était d'avoir toutes les permissions de l'utilisateur sur l'utilisateur que ça soit une permission lui permettant d'avoir accès à un article ou une permission lui permettant d'avoir accès "aux données" (à déterminer) d'un autre utilisateur.

D'ailleurs, sur un article je pense qu'on devrait avoir un unique "owner" ou "creator".

Pour lister l'ensemble des articles d'un utilisateur il faudrait alors faire:

```js
const userIds = user.permissions.filter((p) => p.scope === 'user').map((p) => p.userId).concat([user.id])
const articleIds =  user.permissions.filter((p) => p.scope === 'article').map((p) => p.articleId)
```
```js
db.articles.find({
  $or: [
    { owner: { $in: userIds } },
    { _id: { $in: articleIds } }
  ]
})
```

Si on détaille la requête ci-dessus, on récupère l'ensemble des articles pour lesquels on est owner ainsi que l'ensemble des articles dont le owner est un utilisateur qui nous a donné accès à ses données ainsi que l'ensemble des articles pour lesquels on nous a donné accès.

On peut éventuellement ajouter une liste de "contributors" sur un article si on souhaite "dupliquer" l'information. On pourrait aussi utiliser un "role" pour qualifier les droits accordés à l'utilisateur (par exemple : lecteur, commentateur, éditeur). Actuellement, tout le monde est éditeur mais cela pourrait changer.

```
Article {
  contributors: [
    {
       userId: '1234',
       role: 'reader'
    }
  ]
}
```

Quand on copie un article pour le partager je pense qu'il faut changer le owner/creator.

thom4parisot commented:
Oui j'ai une hésitation sur les permissions, en terme de conception, et de responsabilité (quel domaine gère quoi).

Le fait que la table users stocke la liste des articles, et que les articles stockent la liste des users crée de la confusion.

Dans la table utilisateurs, je mettrais les permissions "de compte" (avoir le droit de lire/écrire dans _tous_ les articles de telle(s) autre(s) personne(s)).

Dans la table articles, je mettrais les permissions "d'article".

Mais en fait je m'en fiche, comme tu dis, c'est l'axe d'entrée dans la base qui m'importe le plus. Si on lit un article, je trouverais ça idéal que la table s'auto-suffise — la session ramène déjà l'utilisateur actif, dont possiblement la liste des users auquels iel a le droit de lire/modifier les articles.

ggrossetie commented:
> Dans la table articles, je mettrais les permissions "d'article".

La permission est forcément rattachée à un utilisateur, c'est une permission que l'on donne pour un article mais on la donne à un utilisateur.

> Mais en fait je m'en fiche, comme tu dis, c'est l'axe d'entrée dans la base qui m'importe le plus. 

Exact.

> Si on lit un article, je trouverais ça idéal que la table s'auto-suffise 

C'est la notion de `owners` actuellement qui sert à lister le créateur de l'article ainsi que les *identifiants* des utilisateurs ayant le droit "éditeur" sur l'article.

D'où ma suggestion de gérer les permissions sur l'utilisateur et de "dupliquer" l'information si nécessaire. À noter qu'on fera quand même des requêtes additionnelles si on stocke uniquement l'identifiant des "contributors" (et même du "owner") sur un article si on souhaite afficher leurs noms.

Il faudra surement optimiser cette partie (plutôt que de laisser faire Mongoose) car je pense que c'est plus efficace de faire une requête qui récupère l'ensemble des "contributors" pour l'ensemble des articles et ensuite de parcourir en mémoire les articles pour faire l'association que de laisser Mongoose faire une ou plusieurs requêtes Mongo par article.

Ça me fait penser qu'actuellement la liste des "contributors" est visible pour le "owner" mais aussi par les "contributor". Autrement dit, si je te partage un article, tu peux savoir que je le partage aussi à Paul et Marie. Je ne sais pas si c'est bien ou pas bien mais avec une distinction claire entre le owner et les contributors on peut faire des règles différentes.

Commit:
Verify user, and populate in a leaner way

Commit:
accepts a string timestamp, as returned per the GraphQL API

thom4parisot commented on graphql/resolvers/nestedModel.js:
```diff
@@ -17,23 +14,21 @@ const Version = require('../models/version');
 
 const getUserById = async (userId) => {
-    const user = await User.findById(userId);
+    const user = await User.findById(userId).lean();
```
https://mongoosejs.com/docs/tutorials/lean.html

Ça retourne la structure de données, et non le "Document" Mongoose.

ggrossetie replied:
Intéressant ça, belle trouvaille 👍🏻

thom4parisot replied:
(ça éviterait tous les contournements basés sur `_doc`)

Commit:
Remove manual document population… and it works just fine

thom4parisot commented on graphql/resolvers/nestedModel.js:
```diff
@@ -1,202 +1,41 @@
+      .populate('tags acquintances')
+      // see https://mongoosejs.com/docs/api/document.html#document_Document-populate
+      // for subdocument population
+      .populate({ path: 'articles', populate: { path: 'owners versions tags' }})
```
Et… avec ça on récupère les infos des sous-documents.

Ça vire tout besoin aux `populateXXX()` et à toutes les requêtes imbriquées.
Je découvre au passage que MongoDB 3.2+ possède un [système de jointure natif](https://docs.mongodb.com/manual/reference/operator/aggregation/lookup/).

Commit:
Credentials screen works

No mutation tested so far

Commit:
Add the ability to remove the sharing with someone

Commit:
Remove an article sharing from the UI

Commit:
Update user password

thom4parisot commented:
![image](https://user-images.githubusercontent.com/138627/146172996-6315a675-3536-4941-92fd-eb44b6e03fa2.png)

Commit:
Grant and revoke account access from the API

Commit:
Grant and revoke account access from the UI

thom4parisot commented:
![image](https://user-images.githubusercontent.com/138627/146189747-a99c62fb-8d3d-4eff-a94b-ffa39615acf9.png)

Commit:
Introduce CredentialsAccountSharing component

ggrossetie commented on graphql/resolvers/articleResolver.js:
```diff
@@ -70,163 +62,163 @@ module.exports = {
-      }
+    //Remove article from owner "user" etc.
+    fetchedArticle.owners.pull(args.to)
+    fetchedUser.articles.pull(args.article)
```
C'est pas redondant d'avoir la liste des articles sur l'utilisateur alors qu'on peut avoir cette information en filtrant les articles avec le owner ?

D'ailleurs est-ce que la liste correspond uniquement aux articles crées par l'utilisateur ou à la liste complète des articles pour lesquels l'utilisateur a des droits ?

thom4parisot replied:
Si c'est carrément redondant. Je ne me suis pas encore attaqué à nettoyer cette partie.

Commit:
Populate shared account articles

Commit:
List articles and display article by Id

Commit:
Shared user is able to autosave

Commit:
Remove nodemon, as pm2 does the job

Commit:
Remove nodemon, as pm2 does the job

Commit:
Remove nodemon, as pm2 does the job

Commit:
Export app calls the GraphQL API, with a super admin token

Commit:
Fix date formatting with new API date output

Commit:
Fetch Article and Version by ID

It also works via the export service

thom4parisot commented on export/src/app.js:
```diff
@@ -33,16 +29,5 @@ app.get('/api/v1/htmlArticle/:id', exportArticleHtml)
-  .catch(err => {
-    console.log('Unable to connect to MongoDB', err)
-  })
+console.log('Listening on http://localhost:%s', listenPort)
```
L'export consomme les données GraphQL. On n'a plus le modèle en doublon. Il n'y a plus de dépendance à MongoDB sur ce service.

thom4parisot commented on export/src/graphql.js:
```diff
@@ -0,0 +1,77 @@
+}
+
+async function getBookById (bookId) {
+  // const book = await Tag.findById(bookId)
```
Y'a ce petit bout là à finir. Je m'en occupe ensuite.

Commit:
List and export Books

Commit:
Stop reloading Books after renaming

Commit:
Make it explicit we are in dev mode

Commit:
Manage tags only if you are the tag owner

thom4parisot commented:
# Tags pour le "owner"

![image](https://user-images.githubusercontent.com/138627/146383416-a3ba54a4-4577-4519-a976-fb4d1604781c.png)

# Tags pour les contributeurs

![image](https://user-images.githubusercontent.com/138627/146383500-7cd4a6c0-1937-4e47-a5a2-f9e7b0571b41.png)

thom4parisot commented:
Distinction article à soi, et article dont on a les droits les collaboration

![image](https://user-images.githubusercontent.com/138627/146387175-d8e36a72-40d3-4af3-bb27-e34a5f52371e.png)

Commit:
You cannot edit tags nor delete an article when you are contributor

Commit:
Restart server when these files change

Commit:
Rename  into , and make it work cross-accounts

Commit:
Remove oncletom from the Store

Commit:
Refactor Article/User methods in

Commit:
Populate user profile via a GraphQL request

This is useful to deal with a unique date format.

GraphQL outputs a date as an integer string

Commit:
Mutualize token generation code

thom4parisot commented:
@Mogztter bon, j'ai l'impression d'avoir nettoyé la majeure partie du code.

Y'aura plus qu'à terminer la migration des permissions côté article, et d'implémenter la nouvelle logique (1 owner, 0 à plusieurs contributors — ou toutes les permissions sous forme de scope côté User).

Commit:
Migrate owners as ower+contributors and expose them in GraphQL

Commit:
Replace owners by owner+contributors in Articles list

ggrossetie committed:
Ajout de l'attribut permissions dans les users

ggrossetie committed:
Mise en conformité du format de createdAt et updatedAt

Commit:
Use owner+contributors for the Write compoonent

Commit:
Use owner+contributors for the Write compoonent

Commit:
Fix article saving

ggrossetie committed:
Ajout de l'apiToken dans la réponse (pour la page credentials)

Commit:
Enable article sharing

Commit:
Include initial userId in findAccountAccessUserIds

Commit:
Include initial userId in findAccountAccessUserIds

ggrossetie committed:
Verifie que l'utilisateur a accès à la version demandée (via l'article)

ggrossetie committed:
Supprime unlinkVersion (on ne comprends pas à quoi ça sert)

Commit:
Remove unused Feather Icons in CreateVersion.jsx

Commit:
Remove mentions to owners, and factorise into Article.{findOneByOwner,findAndPopulateByOwners}

Commit:
Make pm2 watch eventually work

Commit:
article key is actually unused in Version schema…

Commit:
Restore initial version of 'version' GraphQL resolver, as it works

We will address its unrestricted access in a separate issue

thom4parisot commented on graphql/models/version.js:
```diff
@@ -6,10 +6,6 @@ const versionSchema = new Schema({
     type: Schema.Types.ObjectId,
     ref: 'User'
   },
-  article:{
-    type: Schema.Types.ObjectId,
-    ref: 'Article'
-  },
```
Je découvre qu'en fait c'était dans le modèle, mais que la jointure n'était jamais effectuée (pas de référence assignée lors de la création d'une nouvelle version). Ce qui rend l'usage de `version.populate('article')` impossible.

Je supprime pour éviter tout doute/implicite dans le schéma, et l'absence de donnée associée en base.

Commit:
Migrate Account Access

Commit:
Remove token from response

It was probably bugging because of a GraphQL response error

thom4parisot commented:
J'ai ajouté les éléments de migration des partages de compte dans le script `graphql/data/20212114-passwords.js`. J'ai fait le tour de l'application mais j'ai peut-être raté des trucs !

Je commence à gagner en confiance sur le fait qu'on pourrait envisager de déployer cette PR sur stylo-dev dans les jours qui viennent.

antoinentl commented:
Déjà bravo pour ce travail ! Je note ici nos remarques à @Mmellet et à moi. En l'état il faut revoir certains fonctionnements, il va falloir attendre avant de merger.
Je structure un peu nos remarques :

## Export
Tout fonctionne parfaitement, j'ai testé plusieurs scénarios d'export.

## Partage de compte
### Action supplémentaire pour valider
Le bouton qui permet de donner accès/de partager un compte est trop facile d'accès : il faudrait ajouter une popup ou un écran pour confirmer le partage, en reprenant la phrase d'avertissement. Le risque étant que quelqu'un clique dessus presque par inadvertance, alors que cela pourrait avoir de fortes répercussions.

### Ajouter un compte/une adresse mail
Pour l'instant le partage de compte est limité aux adresses/comptes avec qui nous avons déjà partagé un article, ce qui ajoute une étape si nous souhaitons ajouter quelqu'un avec qui nous n'avons jamais partagé d'article. Il faudrait donc remettre le champ permettant d'ajouter manuellement une adresse et ensuite partager le compte.

### Mettre en évidence les comptes partagé
Dans la liste il faudrait mettre en évidence les comptes avec lesquels notre compte est partagé, par exemple en début de liste ou graphiquement plus visible.

## Partage d'article
Actuellement le partage d'un article implique qu'aucune personne qui a accès à l'article n'a plus de _droits_ qu'une autre, typiquement il n'est pas possible pour une personne qui a partagé un article de supprimer l'article _chez la personne avec qui l'article est partagé_. Une piste de solution serait d'ajouter des droits de partage (voir ci-dessous), mais cela pose un problème de fond : si deux personnes ont un accès à un article, il ne faut pas que la première puisse supprimer l'article et ainsi empêcher la seconde de continuer à travailler dessus. Ce serait dramatique dans l'usage que nous avons de Stylo pour des revues, où un·e auteur·rice pourrait se dire "Ho cet article est en édition chez _Revue machin_, je peux supprimer l'article", alors que le partage entraînera la suppression pure et simple de l'article.

### Préciser les droits
Pour remédier à ce problème il faudrait ajouter des options supplémentaires pour le partage :

- céder les droits (si vous supprimez l'article, les personnes avec qui l'article est partagé auront toujours accès à l'article)
- repartage (vous autorisez les personnes avec qui vous avez partagé l'article de le repartager avec les mêmes conditions)

Je me rends compte qu'après des tests hier et ce matin, je ne peux me connecter à stylo-dev, voici l'erreur rencontrée (j'ai essayé de supprimer mes cookies et de retenter, même résultat).
Sous Chrome :

![Capture d’écran de 2022-01-25 20-19-07](https://user-images.githubusercontent.com/6869488/151087641-f69f2c03-6ab2-4456-b977-bd717f756671.png)

La console sous Firefox :

![Capture d’écran de 2022-01-25 20-40-31](https://user-images.githubusercontent.com/6869488/151089012-7a99bee0-ba0b-4a68-84ab-b9d011d53096.png)

ggrossetie commented:
Merci pour ce retour très complet et détaillé !
Je suis en phase avec tes remarques concernant le partage de compte.

Pour le partage d'article, j'ai quelques questions complémentaires.

> Actuellement le partage d'un article implique qu'aucune personne qui a accès à l'article n'a plus de droits qu'une autre, typiquement il n'est pas possible pour une personne qui a partagé un article de supprimer l'article chez la personne avec qui l'article est partagé.

Actuellement, il y a une notion de "owner" (le créateur de l'article) et de contributeurs.
Je pense qu'il est important de comprendre les motivations de la personne qui supprime un article qu'elle a crée et partagé.

"Ho cet article est en édition chez Revue machin, je peux supprimer l'article"

Cela veut dire que l'auteur·rice ne pourra plus consulter ou faire de modifications sur son article même si Revue machin lui fait un retour ? Est-ce que le besoin n'est pas plutôt d'archiver/masquer/ranger l'article ?

Si on veut céder un article, il est déjà possible d'envoyer une copie et de supprimer l'original. On pourrait imaginer un bouton "Céder l'article" qui enchaîne ces deux actions.

Mmellet commented:
En effet cette question de partage est bien complexe ! 

Le problème que souligne @antoinentl  ici est plus un problème pour l'utilisation de Stylo type revue (mais selon moi il pourrait se poser pour d'autres utilisations) : imaginons qu'une auteure partage son article avec la revue Sens public pour publication, les éditeurs et l'auteure travaillent en collaboratif sur l'article, puis l'article est publié ou sur le point de l'être. l'auteure n'a donc plus de raisons de travailler sur son article Stylo, et le supprime (se disant peut-être l'article a été partagé avec la revue, moi je n'en est plus besoin). Sauf que, au vu du fonctionnement du deploy, si l'auteure supprime son article de son compte, cette suppression est en cascade : la revue n'aura plus l'article à publier et je vois mal comment le récupérer. 

En effet avec le deploy, on distingue mieux le share du send. Mais il nous semblait important de pouvoir conserver l'option d'un travail collaboratif ouvert : soit que le partage permette le repartage, permette aussi de concerver une version. 

La notion d'archive que tu mentionne me plaît bien et ça me fait penser à peut-être une alternative : est-ce qu'un tel fonctionnement serait envisageable : mettons qu'on départage l'article d'une personne (si on le départage c'est qu'on considère que le travail collaboratif est terminé)/ ou qu'on le supprime (ce qui veut dire là qu'on considère qu'il n'y a plus de travail à faire), la personne avec qui on a partagé l'article a toujours l'article dans son compte stylo mais en "figé" : ne pouvant pas éditer, ou faire de nouvelle version, mais puis consulter les anciennes versions. 

C'est une proposition qu'il faudra qu'on discute, aussi pour que nous puissions comprendre à quel point un triple système (partage ouvert, partage avec conditions fermées, envoi) serait compliqué à mettre en place.

thom4parisot commented:
La notion d'archivage correspond davantage à l'usage que vous avez de la "fonctionnalité suppression" actuelle. Je vois bien la fonctionnalité Archiver masque l'article de sa liste, tout en donnant la capacité de retourner dessus (via un filtre qui dit "afficher les articles archivés" — un peu comme sur Gmail par exemple).

Un article archivé pour soi, n'est pas un article archivé pour d'autre — c'est une préférence personnelle en fin de compte.

Pour l'instant, on a désactivé la notion de suppression pour un article partagé dont on est l'auteur·ice principale. Ça évite de provoquer le cas que tu décris — où tout le monde perd l'accès à ce document.
